<img src="https://github.com/insaid2018/Term-1/blob/master/Images/INSAID_Full%20Logo.png?raw=true" width="240" height="360" />

# ML1 Project using HousePrice Dataset

## Table of Contents

1. [Problem Statement](#section1)<br>
2. [Data Loading and Description](#section2)
3. [Pandas Profiling](#section3)
4. [Preprocessing](#section4)
5. [EDA](#section5)
6. [Feature Selection using Random Forest](#section6)
7. [Linear Regression using Selected Features](#section7)<br>
    - 7.1 [Linear regression in scikit-learn](#section401)<br>
    - 7.2 [Interpreting Model Coefficients](#section402)<br>
    - 7.3 [Using the Model for Prediction](#section403)<br>
    - 7.4 [Model evaluation using metrics](#section404)<br>
    - 7.5 [Model Evaluation using Rsquared value.](#section404)<br>
8. [Logistic Regression](#section8)<br>

9. [Decision Tree using selected features](#section9)<br>
    - 9.1 [Decision Tree in scikit-learn](#section501)<br>
         - 9.1.1 [Using the Model for Prediction](#randomsampling)<br>    
         - 9.1.2 [Model evaluation using Metrics](#randomsampling)<br>    
         - 9.1.3 [__R Square__](#randomsampling)<br>             
    - 9.2 [Decision Tree with Gridsearch](#section502)<br>
         - 9.2.1 [Model Evaluation using R Square](#randomsampling)<br>
    - 9.3 [Decision Tree with Gridsearch [Hyperparameter Optimization] - Changing Hyperparameters Manually](#section502)<br>
         - 9.3.1 [Model Evaluation using R Square](#randomsampling)<br>
    - 9.4 [Hyperparameter Optimization with Gridsearch -Auto Paramater Optimization](#section502)<br>
         - 9.4.1 [Model Evaluation using R Square](#randomsampling)<br>

10. [Random Forest using selected features](#section10)<br>
    - 10.1 [Random Forest without parameter in scikit-learn](#section501)<br>
         - 10.1.1 [__Model without parameter specification__](#randomsampling)<br>    
         - 10.1.2 [Using the Model for Prediction](#randomsampling)<br>    
         - 10.1.3 [Model evaluation for Random Forest](#randomsampling)<br>             
    - 10.2 [Random Forest - Hyper parameter for RF tuning using RANDOM SEARCH](#section501)<br>
         - 10.2.1 [__Model Evaluation__ using __r2_score__](#randomsampling)<br>    

11. [Comparision of Models](#section11)<br>
    - 10.1 [Model - Linear Regression](#section501)<br>
    - 11.2 [Model - Logistic Regression](#section501)<br>
    - 11.3 [Model - Decision Tree](#section501)<br>
    - 11.4 [Model - Random Forest](#section501)<br>    
12. [Conclusion](#section12)

### 1. Problem Statement

The goal is to __predict house price__ using __Linear Regression, Logistic Regression, Decision Tree and Random Forest__.

### 2. Data Loading and Description

- The dataset consists of the information about people boarding the famous RMS Titanic. Various variables present in the dataset includes data of age, sex, fare, ticket etc. 
- The dataset comprises of __891 observations of 12 columns__. Below is a table showing names of all the columns and their description.

| Column Name   | Description                                               |
| ------------- |:-------------                                            :| 
| PassengerId   | Passenger Identity                                        | 
| Survived      | Whether passenger survived or not                         |  
| Pclass        | Class of ticket                                           | 
| Name          | Name of passenger                                         |   
| Sex           | Sex of passenger                                          |
| Age           | Age of passenger                                          |
| SibSp         | Number of sibling and/or spouse travelling with passenger |
| Parch         | Number of parent and/or children travelling with passenger|
| Ticket        | Ticket number                                             |
| Fare          | Price of ticket                                           |
| Cabin         | Cabin number                                              |
| Embarked      | Port of embarkation                                       |

#### Importing packages                                          

In [3]:
import numpy as np                     

import pandas as pd
pd.set_option('mode.chained_assignment', None)      # To suppress pandas warnings.
pd.set_option('display.max_colwidth', None)         # To display all the data in each column
pd.set_option('display.max_columns', None)          # To display every column of the dataset in head()

import warnings
warnings.filterwarnings('ignore')                   # To suppress all the warnings in the notebook.

In [4]:
import matplotlib.pyplot as plt
%matplotlib inline

import seaborn as sns
sns.set(style='whitegrid', font_scale=1.3, color_codes=True)      # To apply seaborn styles to the plots.

In [5]:
# Making plotly specific imports

# pip install chart-studio

'''
# If you're using an older version of plotly, you might have to import the below modules as well.
from plotly.offline import init_notebook_mode, iplot
import plotly.graph_objs as go
import chart_studio.plotly as py
from plotly import tools
init_notebook_mode(connected=True)
'''

import plotly.graph_objs as go

#### Importing the Dataset

In [6]:
houseprice = pd.read_csv('https://raw.githubusercontent.com/insaid2018/Term-2/master/Projects/house_data.csv')

In [7]:
houseprice.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1         Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
2         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
3         Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
4         Lvl    AllPub       FR2       Gtl      NoRidge       Norm   

  Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0       Norm     1Fam     2Story            7            5       2003   
1       Norm     1Fam     1Story            6            8       1976   
2       Norm     1Fam     2Story            7            5       2001   
3       Norm     1Fam     2Story            7            5       1915   
4       Norm     1Fam     2Story            8            5       2000   

   YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0          2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1          1976     Gable  CompShg     MetalSd     MetalSd       None   
2          2002     Gable  CompShg     VinylSd     VinylSd    BrkFace   
3          1970     Gable  CompShg     Wd Sdng     Wd Shng       None   
4          2000     Gable  CompShg     VinylSd     VinylSd    BrkFace   

   MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
0       196.0        Gd        TA      PConc       Gd       TA           No   
1         0.0        TA        TA     CBlock       Gd       TA           Gd   
2       162.0        Gd        TA      PConc       Gd       TA           Mn   
3         0.0        TA        TA     BrkTil       TA       Gd           No   
4       350.0        Gd        TA      PConc       Gd       TA           Av   

  BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  \
0          GLQ         706          Unf           0        150          856   
1          ALQ         978          Unf           0        284         1262   
2          GLQ         486          Unf           0        434          920   
3          ALQ         216          Unf           0        540          756   
4          GLQ         655          Unf           0        490         1145   

  Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  \
0    GasA        Ex          Y      SBrkr       856       854             0   
1    GasA        Ex          Y      SBrkr      1262         0             0   
2    GasA        Ex          Y      SBrkr       920       866             0   
3    GasA        Gd          Y      SBrkr       961       756             0   
4    GasA        Ex          Y      SBrkr      1145      1053             0   

   GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  \
0       1710             1             0         2         1             3   
1       1262             0             1         2         0             3   
2       1786             1             0         2         1             3   
3       1717             1             0         1         0             3   
4       2198             1             0         2         1             4   

   KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu  \
0             1          Gd             8        Typ           0         NaN   
1             1          TA             6        Typ           1          TA   
2             1          Gd             6        Typ           1          TA   
3             1          Gd            

In [8]:
houseprice.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

In [9]:
houseprice.shape

(1460, 81)

In [10]:
houseprice.describe(include='all')

Id   MSSubClass MSZoning  LotFrontage        LotArea Street  \
count   1460.000000  1460.000000     1460  1201.000000    1460.000000   1460   
unique          NaN          NaN        5          NaN            NaN      2   
top             NaN          NaN       RL          NaN            NaN   Pave   
freq            NaN          NaN     1151          NaN            NaN   1454   
mean     730.500000    56.897260      NaN    70.049958   10516.828082    NaN   
std      421.610009    42.300571      NaN    24.284752    9981.264932    NaN   
min        1.000000    20.000000      NaN    21.000000    1300.000000    NaN   
25%      365.750000    20.000000      NaN    59.000000    7553.500000    NaN   
50%      730.500000    50.000000      NaN    69.000000    9478.500000    NaN   
75%     1095.250000    70.000000      NaN    80.000000   11601.500000    NaN   
max     1460.000000   190.000000      NaN   313.000000  215245.000000    NaN   

       Alley LotShape LandContour Utilities LotConfig LandSlope Neighborhood  \
count     91     1460        1460      1460      1460      1460         1460   
unique     2        4           4         2         5         3           25   
top     Grvl      Reg         Lvl    AllPub    Inside       Gtl        NAmes   
freq      50      925        1311      1459      1052      1382          225   
mean     NaN      NaN         NaN       NaN       NaN       NaN          NaN   
std      NaN      NaN         NaN       NaN       NaN       NaN          NaN   
min      NaN      NaN         NaN       NaN       NaN       NaN          NaN   
25%      NaN      NaN         NaN       NaN       NaN       NaN          NaN   
50%      NaN      NaN         NaN       NaN       NaN       NaN          NaN   
75%      NaN      NaN         NaN       NaN       NaN       NaN          NaN   
max      NaN      NaN         NaN       NaN       NaN       NaN          NaN   

       Condition1 Condition2 BldgType HouseStyle  OverallQual  OverallCond  \
count        1460       1460     1460       1460  1460.000000  1460.000000   
unique          9          8        5          8          NaN          NaN   
top          Norm       Norm     1Fam     1Story          NaN          NaN   
freq         1260       1445     1220        726          NaN          NaN   
mean          NaN        NaN      NaN        NaN     6.099315     5.575342   
std           NaN        NaN      NaN        NaN     1.382997     1.112799   
min           NaN        NaN      NaN        NaN     1.000000     1.000000   
25%           NaN        NaN      NaN        NaN     5.000000     5.000000   
50%           NaN        NaN      NaN        NaN     6.000000     5.000000   
75%           NaN        NaN      NaN        NaN     7.000000     6.000000   
max           NaN        NaN      NaN        NaN    10.000000     9.000000   

          YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd  \
count   1460.000000   1460.000000      1460     1460        1460        1460   
unique          NaN           NaN         6        8          15          16   
top             NaN           NaN     Gable  CompShg     VinylSd     VinylSd   
freq            NaN           NaN      1141     1434         515         504   
mean    1971.267808   1984.865753       NaN      NaN         NaN         NaN   
std       30.202904     20.645407       NaN      NaN         NaN         NaN   
min     1872.000000   1950.000000       NaN      NaN         NaN         NaN   
25%     1954.000000   1967.000000       NaN      NaN         NaN         NaN   
50%     1973.000000   1994.000000       NaN      NaN         NaN         NaN   
75%     2000.000000   2004.000000       NaN      NaN         NaN         NaN   
max     2010.000000   2010.000000       NaN      NaN         NaN         NaN   

       MasVnrType   MasVnrArea ExterQual ExterCond Foundation BsmtQual  \
count        1452  1452.000000      1460      1460       1460     1423   
unique          4          NaN         4         

## 3. Pandas Profiling

In [10]:
import pandas_profiling

In [11]:
profile = houseprice.profile_report(title="House Data Profiling before Data Preprocessing", progress_bar=False, minimal=True)
profile.to_file(output_file="house_data_profiling_before_preprocessing.html")

#### Observations from Profiling

**** SalePrice - the property's sale price in dollars. This is the target variable that you're trying to predict.**** <BR>
MSSubClass: The building class **good data, continous values, value 20 has 536 count**<BR>
MSZoning: The general zoning classification **categorical - 5 values, RL has 1151 count**<BR>
LotFrontage: Linear feet of street connected to property **Good histogram, 17.7% missing value**<BR>
LotArea: Lot size in square feet **good histogram, 0% missing value**<BR>
Street: Type of road access ** Categorical, 2 Distinct, 0% missing value**<BR>
-- Alley: Type of alley access **categorical, 2 distinct, 93.8 % missing value, can be deleted.**<BR>
LotShape: General shape of property **Categorical, 4 distinct value, 0% Missing values**<BR>
LandContour: Flatness of the property **Categorical, 4 distinct value, 0% Missing values**<BR>
Utilities: Type of utilities available **CAtegorical, 2 distinct values, 0% Missing vale**<BR>
LotConfig: Lot configuration **Categorical, 5 distinct values, 0% missing values**<BR>
LandSlope: Slope of property **Categorical, 3 distinct values, 0% missing values**<BR>
Neighborhood: Physical locations within Ames city limits **Categorical,  distinct values, 0% missing values**<BR>
Condition1: Proximity to main road or railroad**Categorical,  distinct values, 0% missing values**<BR>
Condition2: Proximity to main road or railroad (if a second is present)**Categorical,  distinct values, 0% missing values**<BR>
BldgType: Type of dwelling**Categorical,  distinct values, 0% missing values**<BR>
HouseStyle: Style of dwelling**Categorical,  distinct values, 0% missing values**<BR>
OverallQual: Overall material and finish quality **good histogram, 0% missing values**<BR>
OverallCond: Overall condition rating **good histogram, 0% missing values**<BR>
YearBuilt: Original construction date **year column**<BR>
YearRemodAdd: Remodel date **year 1950 has 178 values**<BR>
RoofStyle: Type of roof **Categorical,  distinct values, 0% missing values**<BR>
RoofMatl: Roof material**Categorical,  distinct values, 0% missing values**<BR>
Exterior1st: Exterior covering on house**Categorical,  distinct values, 0% missing values**<BR>
Exterior2nd: Exterior covering on house (if more than one material)**Categorical,  distinct values, 0% missing values**<BR>
MasVnrType: Masonry veneer type **Categorical,  distinct values, 0.5% missing values**<BR>
MasVnrArea: Masonry veneer area in square feet **Right Skewed, Median is 0, mean is 104**<BR>
ExterQual: Exterior material quality **Categorical,  distinct values, 0% missing values**<BR>
ExterCond: Present condition of the material on the exterior **Categorical,  distinct values, 0% missing values**<BR>
Foundation: Type of foundation **Categorical,  distinct values, 0% missing values**<BR>
BsmtQual: Height of the basement**Categorical,  distinct values, 2.5% missing values**<BR>
BsmtCond: General condition of the basement **Categorical,  distinct values, 2.5% missing values**<BR>
BsmtExposure: Walkout or garden level basement walls **Categorical,  distinct values, 2.6% missing values**<BR>
BsmtFinType1: Quality of basement finished area **Categorical,  distinct values, 2.5% missing values**<BR>
BsmtFinSF1: Type 1 finished square feet **Right skewed, 0% missing values.**<BR>
BsmtFinType2: Quality of second finished area (if present)**Categorical,  distinct values, 2.6% missing values**<BR>
BsmtFinSF2: Type 2 finished square feet **Right skewed, 0 % missing values**<BR>
BsmtUnfSF: Unfinished square feet of basement area **right skewed, 0% ,missing values**<BR>
TotalBsmtSF: Total square feet of basement area **Good histogram, 0% missing values**<BR>
Heating: Type of heating **Categorical,  distinct values, 0% missing values**<BR>
HeatingQC: Heating quality and condition **Categorical,  distinct values, 0% missing values**<BR>
CentralAir: Central air conditioning**Boolean,  2 distinct values, 0% missing values**<BR>
Electrical: Electrical system **Categorical,  distinct values, 0.1% missing values**<BR>
1stFlrSF: First Floor square feet **good histogram, 0% missing values**<BR>
2ndFlrSF: Second floor square feet **0 median, 0% missing data**<BR>
LowQualFinSF: Low quality finished square feet (all floors) **0 median, 0% missing data**<BR>
GrLivArea: Above grade (ground) living area square feet **Good histogram, 0% missing values**<BR>
BsmtFullBath: Basement full bathrooms**Categorical,  distinct values, 0% missing values**<BR>
BsmtHalfBath: Basement half bathrooms**Categorical,  distinct values, 0% missing values**<BR>
FullBath: Full bathrooms above grade**Categorical,  distinct values, 0% missing values**<BR>
HalfBath: Half baths above grade**Categorical,  distinct values, 0% missing values**<BR>
Bedroom: Number of bedrooms above basement level **good histogram, 0% missing values**<BR>
Kitchen: Number of kitchens **Categorical,  distinct values, 0% missing values**<BR>
KitchenQual: Kitchen quality**Categorical,  distinct values, 0% missing values**<BR>
TotRmsAbvGrd: Total rooms above grade (does not include bathrooms) **Good histogram, 0% missing values**<BR>
Functional: Home functionality rating **Categorical,  distinct values, 0% missing values**<BR>
Fireplaces: Number of fireplaces **Categorical,  distinct values, 0% missing values**<BR>
-- FireplaceQu: Fireplace quality **Categorical,  distinct values, 47.3% missing values**<BR>
GarageType: Garage location **Categorical,  distinct values, 5.5% missing values**<BR>
GarageYrBlt: Year garage was built **year column, 5.5% missing data**<BR>
GarageFinish: Interior finish of the garage **Categorical,  distinct values, 5.5% missing values**<BR>
GarageCars: Size of garage in car capacity **good histogram, 0% missing values**<BR>
GarageArea: Size of garage in square feet **good histogram, 0% missing data.**<BR>
GarageQual: Garage quality **Categorical,  distinct values, 5.5% missing values**<BR>
GarageCond: Garage condition **Categorical,  distinct values, 5.5% missing values**<BR>
PavedDrive: Paved driveway **Categorical,  distinct values, 0% missing values**<BR>
WoodDeckSF: Wood deck area in square feet **continuous data,  0% missing values, 52% zeroes**<BR>
-- OpenPorchSF: Open porch area in square feet **CONTINUOUS DATA, 44% zero values**<BR>
-- EnclosedPorch: Enclosed porch area in square feet **Continuous data, 85% of zeroes**<BR>
-- 3SsnPorch: Three season porch area in square feet **continuous data, 98.4% zeroes**<BR>
-- ScreenPorch: Screen porch area in square feet **continous data, 92.1%**<BR>
-- PoolArea: Pool area in square feet **Continuous data, 99.5% zeroes**<BR>
-- PoolQC: Pool quality **Categorical,  distinct values, 99.9% missing values**<BR>
-- Fence: Fence quality **Categorical,  distinct values, 88.8% missing values**<BR>
-- MiscFeature: Miscellaneous feature not covered in other categories **Categorical,  distinct values, 96.3% missing values**<BR>
-- MiscVal: $Value of miscellaneous feature **Continuous data, 96.4 %zeroes**<BR>
MoSold: Month Sold **good histogram, 0% missing data**<BR>
YrSold: Year Sold **0% missing data**<BR>
SaleType: Type of sale **Categorical,  distinct values, 0% missing values**<BR>
SaleCondition: Condition of sale **Categorical,  distinct values, 0% missing values**<BR>

#### 3.1 Data plot's before data cleaning and pre processing for understanding of raw data.

##### 3.1.1 Distplot before data pre Processing .

## 4. Preprocessing the data

#### Creating dummies for categorical features

In [11]:
MSZoning = pd.get_dummies(houseprice.MSZoning, prefix='MSZoning').iloc[:, 1:]
Street = pd.get_dummies(houseprice.MSZoning, prefix='Street').iloc[:, 1:]
LotShape = pd.get_dummies(houseprice.MSZoning, prefix='LotShape').iloc[:, 1:]
LandContour = pd.get_dummies(houseprice.MSZoning, prefix='LandContour').iloc[:, 1:]
Utilities = pd.get_dummies(houseprice.MSZoning, prefix='Utilities').iloc[:, 1:]
LotConfig = pd.get_dummies(houseprice.MSZoning, prefix='LotConfig').iloc[:, 1:]
LandSlope = pd.get_dummies(houseprice.MSZoning, prefix='LandSlope').iloc[:, 1:]
Neighborhood = pd.get_dummies(houseprice.MSZoning, prefix='Neighborhood').iloc[:, 1:]
Condition1 = pd.get_dummies(houseprice.MSZoning, prefix='Condition1').iloc[:, 1:]
BldgType = pd.get_dummies(houseprice.MSZoning, prefix='BldgType').iloc[:, 1:]
HouseStyle = pd.get_dummies(houseprice.MSZoning, prefix='HouseStyle').iloc[:, 1:]
RoofStyle = pd.get_dummies(houseprice.MSZoning, prefix='RoofStyle').iloc[:, 1:]
RoofMatl = pd.get_dummies(houseprice.MSZoning, prefix='RoofMatl').iloc[:, 1:]
Exterior1st = pd.get_dummies(houseprice.MSZoning, prefix='Exterior1st').iloc[:, 1:]
MasVnrType = pd.get_dummies(houseprice.MSZoning, prefix='MasVnrType').iloc[:, 1:]
Alley = pd.get_dummies(houseprice.Alley, prefix='Alley').iloc[:, 1:]
Condition2 = pd.get_dummies(houseprice.Condition2, prefix='Condition2').iloc[:, 1:]
Exterior2nd = pd.get_dummies(houseprice.Exterior2nd, prefix='Exterior2nd').iloc[:, 1:]
ExterQual = pd.get_dummies(houseprice.ExterQual, prefix='ExterQual').iloc[:, 1:]
ExterCond = pd.get_dummies(houseprice.ExterCond, prefix='ExterCond').iloc[:, 1:]
Foundation = pd.get_dummies(houseprice.Foundation, prefix='Foundation').iloc[:, 1:]
BsmtQual = pd.get_dummies(houseprice.BsmtQual, prefix='BsmtQual').iloc[:, 1:]
BsmtCond = pd.get_dummies(houseprice.BsmtCond, prefix='BsmtCond').iloc[:, 1:]
BsmtExposure = pd.get_dummies(houseprice.BsmtExposure, prefix='BsmtExposure').iloc[:, 1:]
BsmtFinType1 = pd.get_dummies(houseprice.BsmtFinType1, prefix='BsmtFinType1').iloc[:, 1:]
BsmtFinType2 = pd.get_dummies(houseprice.BsmtFinType2, prefix='BsmtFinType2').iloc[:, 1:]
Heating = pd.get_dummies(houseprice.Heating, prefix='Heating').iloc[:, 1:]
HeatingQC = pd.get_dummies(houseprice.HeatingQC, prefix='HeatingQC').iloc[:, 1:]
CentralAir = pd.get_dummies(houseprice.CentralAir, prefix='CentralAir').iloc[:, 1:]
Electrical = pd.get_dummies(houseprice.Electrical, prefix='Electrical').iloc[:, 1:]
KitchenQual = pd.get_dummies(houseprice.KitchenQual, prefix='KitchenQual').iloc[:, 1:]
Functional = pd.get_dummies(houseprice.Functional, prefix='Functional').iloc[:, 1:]
FireplaceQu = pd.get_dummies(houseprice.FireplaceQu, prefix='FireplaceQu').iloc[:, 1:]
GarageType = pd.get_dummies(houseprice.GarageType, prefix='GarageType').iloc[:, 1:]
GarageFinish = pd.get_dummies(houseprice.GarageFinish, prefix='GarageFinish').iloc[:, 1:]
GarageQual = pd.get_dummies(houseprice.GarageQual, prefix='GarageQual').iloc[:, 1:]
GarageCond = pd.get_dummies(houseprice.GarageCond, prefix='GarageCond').iloc[:, 1:]
PavedDrive = pd.get_dummies(houseprice.PavedDrive, prefix='PavedDrive').iloc[:, 1:]
PoolQC = pd.get_dummies(houseprice.PoolQC, prefix='PoolQC').iloc[:, 1:]
Fence = pd.get_dummies(houseprice.Fence, prefix='Fence').iloc[:, 1:]
MiscFeature = pd.get_dummies(houseprice.MiscFeature, prefix='MiscFeature').iloc[:, 1:]
SaleType = pd.get_dummies(houseprice.SaleType, prefix='SaleType').iloc[:, 1:]
SaleCondition = pd.get_dummies(houseprice.SaleCondition, prefix='SaleCondition').iloc[:, 1:]

In [12]:
houseprice = pd.concat([houseprice, MSZoning], axis=1)
houseprice = pd.concat([houseprice, Street], axis=1)
houseprice = pd.concat([houseprice, LotShape], axis=1)
houseprice = pd.concat([houseprice, LandContour], axis=1)
houseprice = pd.concat([houseprice, Utilities], axis=1)
houseprice = pd.concat([houseprice, LotConfig], axis=1)
houseprice = pd.concat([houseprice, LandSlope], axis=1)
houseprice = pd.concat([houseprice, Neighborhood], axis=1)
houseprice = pd.concat([houseprice, Condition1], axis=1)
houseprice = pd.concat([houseprice, BldgType], axis=1)
houseprice = pd.concat([houseprice, HouseStyle], axis=1)
houseprice = pd.concat([houseprice, RoofStyle], axis=1)
houseprice = pd.concat([houseprice, RoofMatl], axis=1)
houseprice = pd.concat([houseprice, Exterior1st], axis=1)
houseprice = pd.concat([houseprice, MasVnrType], axis=1)
houseprice = pd.concat([houseprice, Alley], axis=1)
houseprice = pd.concat([houseprice, Condition2], axis=1)
houseprice = pd.concat([houseprice, Exterior2nd], axis=1)
houseprice = pd.concat([houseprice, ExterQual], axis=1)
houseprice = pd.concat([houseprice, ExterCond], axis=1)
houseprice = pd.concat([houseprice, Foundation], axis=1)
houseprice = pd.concat([houseprice, BsmtQual], axis=1)
houseprice = pd.concat([houseprice, BsmtCond], axis=1)
houseprice = pd.concat([houseprice, BsmtExposure], axis=1)
houseprice = pd.concat([houseprice, BsmtFinType1], axis=1)
houseprice = pd.concat([houseprice, BsmtFinType2], axis=1)
houseprice = pd.concat([houseprice, Heating], axis=1)
houseprice = pd.concat([houseprice, HeatingQC], axis=1)
houseprice = pd.concat([houseprice, CentralAir], axis=1)
houseprice = pd.concat([houseprice, Electrical], axis=1)
houseprice = pd.concat([houseprice, KitchenQual], axis=1)
houseprice = pd.concat([houseprice, Functional], axis=1)
houseprice = pd.concat([houseprice, FireplaceQu], axis=1)
houseprice = pd.concat([houseprice, GarageType], axis=1)
houseprice = pd.concat([houseprice, GarageFinish], axis=1)
houseprice = pd.concat([houseprice, GarageQual], axis=1)
houseprice = pd.concat([houseprice, GarageCond], axis=1)
houseprice = pd.concat([houseprice, PavedDrive], axis=1)
houseprice = pd.concat([houseprice, PoolQC], axis=1)
houseprice = pd.concat([houseprice, Fence], axis=1)
houseprice = pd.concat([houseprice, MiscFeature], axis=1)
houseprice = pd.concat([houseprice, SaleType], axis=1)
houseprice = pd.concat([houseprice, SaleCondition], axis=1)

In [13]:
houseprice.drop(['MSZoning'], axis = 1, inplace = True)
houseprice.drop(['Street'], axis = 1, inplace = True)
houseprice.drop(['LotShape'], axis = 1, inplace = True)
houseprice.drop(['LandContour'], axis = 1, inplace = True)
houseprice.drop(['Utilities'], axis = 1, inplace = True)
houseprice.drop(['LotConfig'], axis = 1, inplace = True)
houseprice.drop(['LandSlope'], axis = 1, inplace = True)
houseprice.drop(['Neighborhood'], axis = 1, inplace = True)
houseprice.drop(['Condition1'], axis = 1, inplace = True)
houseprice.drop(['BldgType'], axis = 1, inplace = True)
houseprice.drop(['HouseStyle'], axis = 1, inplace = True)
houseprice.drop(['RoofStyle'], axis = 1, inplace = True)
houseprice.drop(['RoofMatl'], axis = 1, inplace = True)
houseprice.drop(['Exterior1st'], axis = 1, inplace = True)
houseprice.drop(['MasVnrType'], axis = 1, inplace = True)
houseprice.drop(['Alley'], axis = 1, inplace = True)
houseprice.drop(['Condition2'], axis = 1, inplace = True)
houseprice.drop(['Exterior2nd'], axis = 1, inplace = True)
houseprice.drop(['ExterQual'], axis = 1, inplace = True)
houseprice.drop(['ExterCond'], axis = 1, inplace = True)
houseprice.drop(['Foundation'], axis = 1, inplace = True)
houseprice.drop(['BsmtQual'], axis = 1, inplace = True)
houseprice.drop(['BsmtCond'], axis = 1, inplace = True)
houseprice.drop(['BsmtExposure'], axis = 1, inplace = True)
houseprice.drop(['BsmtFinType1'], axis = 1, inplace = True)
houseprice.drop(['BsmtFinType2'], axis = 1, inplace = True)
houseprice.drop(['Heating'], axis = 1, inplace = True)
houseprice.drop(['HeatingQC'], axis = 1, inplace = True)
houseprice.drop(['CentralAir'], axis = 1, inplace = True)
houseprice.drop(['Electrical'], axis = 1, inplace = True)
houseprice.drop(['KitchenQual'], axis = 1, inplace = True)
houseprice.drop(['Functional'], axis = 1, inplace = True)
houseprice.drop(['FireplaceQu'], axis = 1, inplace = True)
houseprice.drop(['GarageType'], axis = 1, inplace = True)
houseprice.drop(['GarageFinish'], axis = 1, inplace = True)
houseprice.drop(['GarageQual'], axis = 1, inplace = True)
houseprice.drop(['GarageCond'], axis = 1, inplace = True)
houseprice.drop(['PavedDrive'], axis = 1, inplace = True)
houseprice.drop(['PoolQC'], axis = 1, inplace = True)
houseprice.drop(['Fence'], axis = 1, inplace = True)
houseprice.drop(['MiscFeature'], axis = 1, inplace = True)
houseprice.drop(['SaleType'], axis = 1, inplace = True)
houseprice.drop(['SaleCondition'], axis = 1, inplace = True)
houseprice

Id  MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  \
0        1          60         65.0     8450            7            5   
1        2          20         80.0     9600            6            8   
2        3          60         68.0    11250            7            5   
3        4          70         60.0     9550            7            5   
4        5          60         84.0    14260            8            5   
...    ...         ...          ...      ...          ...          ...   
1455  1456          60         62.0     7917            6            5   
1456  1457          20         85.0    13175            6            6   
1457  1458          70         66.0     9042            7            9   
1458  1459          20         68.0     9717            5            6   
1459  1460          20         75.0     9937            5            6   

      YearBuilt  YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  \
0          2003          2003       196.0         706           0        150   
1          1976          1976         0.0         978           0        284   
2          2001          2002       162.0         486           0        434   
3          1915          1970         0.0         216           0        540   
4          2000          2000       350.0         655           0        490   
...         ...           ...         ...         ...         ...        ...   
1455       1999          2000         0.0           0           0        953   
1456       1978          1988       119.0         790         163        589   
1457       1941          2006         0.0         275           0        877   
1458       1950          1996         0.0          49        1029          0   
1459       1965          1965         0.0         830         290        136   

      TotalBsmtSF  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  \
0             856       856       854             0       1710             1   
1            1262      1262         0             0       1262             0   
2             920       920       866             0       1786             1   
3             756       961       756             0       1717             1   
4            1145      1145      1053             0       2198             1   
...           ...       ...       ...           ...        ...           ...   
1455          953       953       694             0       1647             0   
1456         1542      2073         0             0       2073             1   
1457         1152      1188      1152             0       2340             0   
1458         1078      1078         0             0       1078             1   
1459         1256      1256         0             0       1256             1   

      BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  \
0                0         2         1             3             1   
1                1         2         0             3             1   
2                0         2         1             3             1   
3                0         1         0             3             1   
4                0         2         1             4             1   
...            ...       ...       ...           ...           ...   
1455             0         2         1             3             1   
1456             0         2         0             3             1   
1457             0         2         0             4             1   
1458             0         1         0             2             1   
1459             0         1         1             3             1   

      TotRmsAbvGrd  Fireplaces  GarageYrBlt  GarageCars  GarageArea  \
0                8           0       2003.0           2         548   
1                6           1       1976.0           2         460   
2                6           1       2001.0           2         608   
3                7           1       1998.0           3         642   
4

In [14]:
#Use these lines of code to print out the column names which contain at least one null value:

for column in houseprice:
    if houseprice[column].isnull().any():
       print('{0} has {1} null values'.format(column, houseprice[column].isnull().sum()))

LotFrontage has 259 null values
MasVnrArea has 8 null values
GarageYrBlt has 81 null values


##### Details of null values in Dataset
LotFrontage has 259 null values <br>
MasVnrArea has 8 null values <br>
GarageYrBlt has 81 null values <br>

#### Replace null or NA with mean of the column

In [15]:
# Replace value NA with mean of the column

houseprice['LotFrontage'].fillna((houseprice['LotFrontage'].mean()), inplace=True)
houseprice['MasVnrArea'].fillna((houseprice['MasVnrArea'].mean()), inplace=True)
houseprice['GarageYrBlt'].fillna((houseprice['GarageYrBlt'].mean()), inplace=True)

## 5. EDA

## 6. Feature Selection using Random Forest

In [16]:
import pandas as pd
from sklearn.ensemble.forest import RandomForestClassifier
from sklearn.feature_selection import SelectFromModel
from sklearn.model_selection import train_test_split

In [17]:
features = ['MSSubClass', 'LotFrontage', 'LotArea', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd', 'MasVnrArea', 'BsmtFinSF1', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd', 'Fireplaces', 'GarageYrBlt', 'GarageCars', 'GarageArea', 'WoodDeckSF', 'OpenPorchSF', 'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea', 'MiscVal', 'MoSold', 'YrSold', 'MSZoning_FV', 'MSZoning_RH', 'MSZoning_RL', 'MSZoning_RM', 'Street_FV', 'Street_RH', 'Street_RL', 'Street_RM', 'LotShape_FV', 'LotShape_RH', 'LotShape_RL', 'LotShape_RM', 'LandContour_FV', 'LandContour_RH', 'LandContour_RL', 'LandContour_RM', 'Utilities_FV', 'Utilities_RH', 'Utilities_RL', 'Utilities_RM', 'LotConfig_FV', 'LotConfig_RH', 'LotConfig_RL', 'LotConfig_RM', 'LandSlope_FV', 'LandSlope_RH', 'LandSlope_RL', 'LandSlope_RM', 'Neighborhood_FV', 'Neighborhood_RH', 'Neighborhood_RL', 'Neighborhood_RM', 'Condition1_FV', 'Condition1_RH', 'Condition1_RL', 'Condition1_RM', 'BldgType_FV', 'BldgType_RH', 'BldgType_RL', 'BldgType_RM', 'HouseStyle_FV', 'HouseStyle_RH', 'HouseStyle_RL', 'HouseStyle_RM', 'RoofStyle_FV', 'RoofStyle_RH', 'RoofStyle_RL', 'RoofStyle_RM', 'RoofMatl_FV', 'RoofMatl_RH', 'RoofMatl_RL', 'RoofMatl_RM', 'Exterior1st_FV', 'Exterior1st_RH', 'Exterior1st_RL', 'Exterior1st_RM', 'MasVnrType_FV', 'MasVnrType_RH', 'MasVnrType_RL', 'MasVnrType_RM', 'Alley_Pave', 'Condition2_Feedr', 'Condition2_Norm', 'Condition2_PosA', 'Condition2_PosN', 'Condition2_RRAe', 'Condition2_RRAn', 'Condition2_RRNn', 'Exterior2nd_AsphShn', 'Exterior2nd_Brk Cmn', 'Exterior2nd_BrkFace', 'Exterior2nd_CBlock', 'Exterior2nd_CmentBd', 'Exterior2nd_HdBoard', 'Exterior2nd_ImStucc', 'Exterior2nd_MetalSd', 'Exterior2nd_Other', 'Exterior2nd_Plywood', 'Exterior2nd_Stone', 'Exterior2nd_Stucco', 'Exterior2nd_VinylSd', 'Exterior2nd_Wd Sdng', 'Exterior2nd_Wd Shng', 'ExterQual_Fa', 'ExterQual_Gd', 'ExterQual_TA', 'ExterCond_Fa', 'ExterCond_Gd', 'ExterCond_Po', 'ExterCond_TA', 'Foundation_CBlock', 'Foundation_PConc', 'Foundation_Slab', 'Foundation_Stone', 'Foundation_Wood', 'BsmtQual_Fa', 'BsmtQual_Gd', 'BsmtQual_TA', 'BsmtCond_Gd', 'BsmtCond_Po', 'BsmtCond_TA', 'BsmtExposure_Gd', 'BsmtExposure_Mn', 'BsmtExposure_No', 'BsmtFinType1_BLQ', 'BsmtFinType1_GLQ', 'BsmtFinType1_LwQ', 'BsmtFinType1_Rec', 'BsmtFinType1_Unf', 'BsmtFinType2_BLQ', 'BsmtFinType2_GLQ', 'BsmtFinType2_LwQ', 'BsmtFinType2_Rec', 'BsmtFinType2_Unf', 'Heating_GasA', 'Heating_GasW', 'Heating_Grav', 'Heating_OthW', 'Heating_Wall', 'HeatingQC_Fa', 'HeatingQC_Gd', 'HeatingQC_Po', 'HeatingQC_TA', 'CentralAir_Y', 'Electrical_FuseF', 'Electrical_FuseP', 'Electrical_Mix', 'Electrical_SBrkr', 'KitchenQual_Fa', 'KitchenQual_Gd', 'KitchenQual_TA', 'Functional_Maj2', 'Functional_Min1', 'Functional_Min2', 'Functional_Mod', 'Functional_Sev', 'Functional_Typ', 'FireplaceQu_Fa', 'FireplaceQu_Gd', 'FireplaceQu_Po', 'FireplaceQu_TA', 'GarageType_Attchd', 'GarageType_Basment', 'GarageType_BuiltIn', 'GarageType_CarPort', 'GarageType_Detchd', 'GarageFinish_RFn', 'GarageFinish_Unf', 'GarageQual_Fa', 'GarageQual_Gd', 'GarageQual_Po', 'GarageQual_TA', 'GarageCond_Fa', 'GarageCond_Gd', 'GarageCond_Po', 'GarageCond_TA', 'PavedDrive_P', 'PavedDrive_Y', 'PoolQC_Fa', 'PoolQC_Gd', 'Fence_GdWo', 'Fence_MnPrv', 'Fence_MnWw', 'MiscFeature_Othr', 'MiscFeature_Shed', 'MiscFeature_TenC', 'SaleType_CWD', 'SaleType_Con', 'SaleType_ConLD', 'SaleType_ConLI', 'SaleType_ConLw', 'SaleType_New', 'SaleType_Oth', 'SaleType_WD', 'SaleCondition_AdjLand', 'SaleCondition_Alloca', 'SaleCondition_Family', 'SaleCondition_Normal', 'SaleCondition_Partial']
# create a Python list of feature names
target = ['SalePrice']                                     # Define the target variable
houseprice[features].shape

(1460, 215)

In [18]:
X_train, X_test, y_train, y_test = train_test_split(houseprice[features], houseprice[target], test_size=0.3, random_state=0)

In [19]:
print(X_train.shape)
print(y_train.shape)

(1022, 215)
(1022, 1)


In [20]:
sel = SelectFromModel(RandomForestClassifier(n_estimators = 100))
sel.fit(X_train, y_train)

SelectFromModel(estimator=RandomForestClassifier())

In [23]:
selectedfeatures= X_train.columns[(sel.get_support())]
len(selectedfeatures)

58

In [24]:
print(selectedfeatures)

Index(['MSSubClass', 'LotFrontage', 'LotArea', 'OverallQual', 'OverallCond',
       'YearBuilt', 'YearRemodAdd', 'MasVnrArea', 'BsmtFinSF1', 'BsmtFinSF2',
       'BsmtUnfSF', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'GrLivArea',
       'BsmtFullBath', 'FullBath', 'HalfBath', 'BedroomAbvGr', 'TotRmsAbvGrd',
       'Fireplaces', 'GarageYrBlt', 'GarageCars', 'GarageArea', 'WoodDeckSF',
       'OpenPorchSF', 'EnclosedPorch', 'ScreenPorch', 'MoSold', 'YrSold',
       'Exterior2nd_HdBoard', 'Exterior2nd_MetalSd', 'Exterior2nd_Plywood',
       'Exterior2nd_VinylSd', 'Exterior2nd_Wd Sdng', 'ExterQual_Gd',
       'ExterQual_TA', 'ExterCond_TA', 'Foundation_CBlock', 'Foundation_PConc',
       'BsmtQual_Gd', 'BsmtQual_TA', 'BsmtExposure_Mn', 'BsmtExposure_No',
       'BsmtFinType1_GLQ', 'BsmtFinType1_Unf', 'HeatingQC_Gd', 'HeatingQC_TA',
       'KitchenQual_Gd', 'KitchenQual_TA', 'FireplaceQu_Gd', 'FireplaceQu_TA',
       'GarageType_Attchd', 'GarageType_Detchd', 'GarageFinish_RFn',
       'GarageF

#### Feature selected using Random Forest:
['MSSubClass', 'LotFrontage', 'LotArea', 'OverallQual', 'OverallCond',
       'YearBuilt', 'YearRemodAdd', 'MasVnrArea', 'BsmtFinSF1', 'BsmtFinSF2',
       'BsmtUnfSF', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'GrLivArea',
       'BsmtFullBath', 'FullBath', 'HalfBath', 'BedroomAbvGr', 'TotRmsAbvGrd',
       'Fireplaces', 'GarageYrBlt', 'GarageCars', 'GarageArea', 'WoodDeckSF',
       'OpenPorchSF', 'EnclosedPorch', 'ScreenPorch', 'MoSold', 'YrSold',
       'Exterior2nd_HdBoard', 'Exterior2nd_MetalSd', 'Exterior2nd_Plywood',
       'Exterior2nd_VinylSd', 'Exterior2nd_Wd Sdng', 'ExterQual_Gd',
       'ExterQual_TA', 'ExterCond_TA', 'Foundation_CBlock', 'Foundation_PConc',
       'BsmtQual_Gd', 'BsmtQual_TA', 'BsmtExposure_Mn', 'BsmtExposure_No',
       'BsmtFinType1_BLQ', 'BsmtFinType1_GLQ', 'BsmtFinType1_Unf',
       'HeatingQC_Gd', 'HeatingQC_TA', 'KitchenQual_Gd', 'KitchenQual_TA',
       'FireplaceQu_Gd', 'FireplaceQu_TA', 'GarageType_Attchd',
       'GarageType_Detchd', 'GarageFinish_RFn', 'GarageFinish_Unf',
       'SaleType_WD', 'SaleCondition_Normal']

## 7. Linear Regression using selected features

In [25]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(houseprice[selectedfeatures], houseprice[target], test_size=0.2, random_state=0)

In [26]:
print('Train cases as below')
print('X_train shape: ',X_train.shape)
print('y_train shape: ',y_train.shape)
print('\nTest cases as below')
print('X_test shape: ',X_test.shape)
print('y_test shape: ',y_test.shape)

Train cases as below
X_train shape:  (1168, 58)
y_train shape:  (1168, 1)

Test cases as below
X_test shape:  (292, 58)
y_test shape:  (292, 1)


In [27]:
X_train.head()

MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
618          20    90.000000    11694            9            5       2007   
870          20    60.000000     6600            5            5       1962   
92           30    80.000000    13360            5            7       1921   
817          20    70.049958    13265            8            5       2002   
302          20   118.000000    13704            7            5       2001   

     YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  \
618          2007       452.0          48           0       1774         1822   
870          1962         0.0           0           0        894          894   
92           2006         0.0         713           0        163          876   
817          2002       148.0        1218           0        350         1568   
302          2002       150.0           0           0       1541         1541   

     1stFlrSF  2ndFlrSF  GrLivArea  BsmtFullBath  FullBath  HalfBath  \
618      1828         0       1828             0         2         0   
870       894         0        894             0         1         0   
92        964         0        964             1         1         0   
817      1689         0       1689             1         2         0   
302      1541         0       1541             0         2         0   

     BedroomAbvGr  TotRmsAbvGrd  Fireplaces  GarageYrBlt  GarageCars  \
618             3             9           1       2007.0           3   
870             2             5           0       1962.0           1   
92              2             5           0       1921.0           2   
817             3             7           2       2002.0           3   
302             3             6           1       2001.0           3   

     GarageArea  WoodDeckSF  OpenPorchSF  EnclosedPorch  ScreenPorch  MoSold  \
618         774           0          108              0          260       7   
870         308           0            0              0            0       8   
92          432           0            0             44            0       8   
817         857         150           59              0            0       7   
302         843         468           81              0            0       1   

     YrSold  Exterior2nd_HdBoard  Exterior2nd_MetalSd  Exterior2nd_Plywood  \
618    2007                    0                    0                    0   
870    2009                    0                    1                    0   
92     2009                    0                    0                    0   
817    2008                    0                    0                    0   
302    2006                    0                    0                    0   

     Exterior2nd_VinylSd  Exterior2nd_Wd Sdng  ExterQual_Gd  ExterQual_TA  \
618                    0                    0             0             0   
870                    0                    0             0             1   
92                     0                    1             0             1   
817                    0                    0             1             0   
302                    1                    0             1             0   

     ExterCond_TA  Foundation_CBlock  Foundation_PConc  BsmtQual_Gd  \
618             1                  0                 1            0   
870             1                  1                 0            0   
92              0                  0                 0            1   
817             1                  0                 1            1   
302             1                  0                 1            1   

     BsmtQual_TA  BsmtExposure_Mn  BsmtExposure_No  BsmtFinType1_GLQ  \
618            0                0                0                 1   
870            1                0                1                 0   
92             0                0                1                 0   
817            0                0                1     

In [28]:
y_train.head()

SalePrice
618     314813
870     109500
92      163500
817     271000
302     205000

In [29]:
X_test.head()

MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
529          20    70.049958    32668            6            3       1957   
491          50    79.000000     9490            6            7       1941   
459          50    70.049958     7015            5            4       1950   
279          60    83.000000    10005            7            5       1977   
655         160    21.000000     1680            6            5       1971   

     YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  \
529          1975  103.685262        1219           0        816         2035   
491          1950    0.000000         403         165        238          806   
459          1950  161.000000         185           0        524          709   
279          1977  299.000000         392           0        768         1160   
655          1971  381.000000           0           0        525          525   

     1stFlrSF  2ndFlrSF  GrLivArea  BsmtFullBath  FullBath  HalfBath  \
529      2515         0       2515             1         3         0   
491       958       620       1578             1         1         0   
459       979       224       1203             1         1         0   
279      1156       866       2022             0         2         1   
655       525       567       1092             0         1         1   

     BedroomAbvGr  TotRmsAbvGrd  Fireplaces  GarageYrBlt  GarageCars  \
529             4             9           2       1975.0           2   
491             3             5           2       1941.0           1   
459             3             5           1       1950.0           1   
279             4             8           1       1977.0           2   
655             3             6           0       1971.0           1   

     GarageArea  WoodDeckSF  OpenPorchSF  EnclosedPorch  ScreenPorch  MoSold  \
529         484           0            0            200            0       3   
491         240           0            0             32            0       8   
459         352           0            0            248            0       7   
279         505         288          117              0            0       3   
655         264           0            0              0            0       3   

     YrSold  Exterior2nd_HdBoard  Exterior2nd_MetalSd  Exterior2nd_Plywood  \
529    2007                    0                    0                    0   
491    2006                    0                    0                    0   
459    2009                    0                    1                    0   
279    2008                    0                    0                    1   
655    2010                    0                    0                    0   

     Exterior2nd_VinylSd  Exterior2nd_Wd Sdng  ExterQual_Gd  ExterQual_TA  \
529                    0                    0             1             0   
491                    0                    1             0             1   
459                    0                    0             0             1   
279                    0                    0             0             1   
655                    0                    0             0             1   

     ExterCond_TA  Foundation_CBlock  Foundation_PConc  BsmtQual_Gd  \
529             1                  0                 1            0   
491             1                  1                 0            0   
459             1                  1                 0            0   
279             1                  1                 0            1   
655             1                  1                 0            0   

     BsmtQual_TA  BsmtExposure_Mn  BsmtExposure_No  BsmtFinType1_GLQ  \
529            1                0                1                 0   
491            1                0                1                 0   
459            1                0                1                 0   
279            0                0                1     

In [30]:
y_test.head()

SalePrice
529     200624
491     133000
459     110000
279     192000
655      88000

#### 7.1 Linear regression in scikit-learn

To apply any machine learning algorithm on your dataset, basically there are 4 steps:
1. Load the algorithm
2. Instantiate and Fit the model to the training dataset
3. Prediction on the test set
4. Evaluate - Calculate RMSE and R square


In [31]:
#Instantiating the model
from sklearn.linear_model import LinearRegression
lr_model = LinearRegression(fit_intercept=True)

In [32]:
lr_model.fit(X_train, y_train)


LinearRegression()

#### 7.2 Interpreting Model Coefficients

In [34]:
print('Intercept:',lr_model.intercept_)          # print the intercept 
print('Coefficients:',lr_model.coef_)  

Intercept: [719663.70373784]
Coefficients: [[-1.53706581e+02  8.90604782e+01  4.57464908e-01  1.24904893e+04
   7.59842724e+03  3.19863702e+02  1.19499894e+02  2.42625610e+01
   1.59060859e+01  9.05272777e+00  1.40849848e+00  2.63673122e+01
   1.01020736e+01  2.31762816e+01  2.88700159e+01  2.25953124e+03
   4.07507865e+03  3.87974366e+03 -5.16266410e+03  1.50155204e+03
   9.05434181e+03 -1.63737740e+01  3.16097278e+03  1.75857892e+01
   1.97094581e+01 -7.36860248e+00  2.76170543e+00  2.53707915e+01
  -4.36077770e+02 -7.73112479e+02 -4.61387537e+03 -8.63839218e+02
  -4.38107559e+03 -3.26230608e+03 -1.31087039e+03 -1.22523898e+04
  -1.73772080e+04  2.89364032e+02 -1.99070462e+03  5.35138307e+03
  -2.17735062e+04 -1.80417663e+04 -6.37166973e+03 -8.40949781e+03
   4.88541893e+03 -1.55644692e+03 -2.73962251e+03 -1.32106340e+03
  -1.78509759e+04 -1.76019880e+04 -4.43482675e+03 -5.66322799e+03
  -1.91798218e+03 -5.88016451e+02 -2.57259403e+03 -2.26577782e+03
  -1.06238198e+04  2.96979733e+03

In [35]:
X_train.columns

Index(['MSSubClass', 'LotFrontage', 'LotArea', 'OverallQual', 'OverallCond',
       'YearBuilt', 'YearRemodAdd', 'MasVnrArea', 'BsmtFinSF1', 'BsmtFinSF2',
       'BsmtUnfSF', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'GrLivArea',
       'BsmtFullBath', 'FullBath', 'HalfBath', 'BedroomAbvGr', 'TotRmsAbvGrd',
       'Fireplaces', 'GarageYrBlt', 'GarageCars', 'GarageArea', 'WoodDeckSF',
       'OpenPorchSF', 'EnclosedPorch', 'ScreenPorch', 'MoSold', 'YrSold',
       'Exterior2nd_HdBoard', 'Exterior2nd_MetalSd', 'Exterior2nd_Plywood',
       'Exterior2nd_VinylSd', 'Exterior2nd_Wd Sdng', 'ExterQual_Gd',
       'ExterQual_TA', 'ExterCond_TA', 'Foundation_CBlock', 'Foundation_PConc',
       'BsmtQual_Gd', 'BsmtQual_TA', 'BsmtExposure_Mn', 'BsmtExposure_No',
       'BsmtFinType1_GLQ', 'BsmtFinType1_Unf', 'HeatingQC_Gd', 'HeatingQC_TA',
       'KitchenQual_Gd', 'KitchenQual_TA', 'FireplaceQu_Gd', 'FireplaceQu_TA',
       'GarageType_Attchd', 'GarageType_Detchd', 'GarageFinish_RFn',
       'GarageF

In [36]:
(lr_model.coef_).T

array([[-1.53706581e+02],
       [ 8.90604782e+01],
       [ 4.57464908e-01],
       [ 1.24904893e+04],
       [ 7.59842724e+03],
       [ 3.19863702e+02],
       [ 1.19499894e+02],
       [ 2.42625610e+01],
       [ 1.59060859e+01],
       [ 9.05272777e+00],
       [ 1.40849848e+00],
       [ 2.63673122e+01],
       [ 1.01020736e+01],
       [ 2.31762816e+01],
       [ 2.88700159e+01],
       [ 2.25953124e+03],
       [ 4.07507865e+03],
       [ 3.87974366e+03],
       [-5.16266410e+03],
       [ 1.50155204e+03],
       [ 9.05434181e+03],
       [-1.63737740e+01],
       [ 3.16097278e+03],
       [ 1.75857892e+01],
       [ 1.97094581e+01],
       [-7.36860248e+00],
       [ 2.76170543e+00],
       [ 2.53707915e+01],
       [-4.36077770e+02],
       [-7.73112479e+02],
       [-4.61387537e+03],
       [-8.63839218e+02],
       [-4.38107559e+03],
       [-3.26230608e+03],
       [-1.31087039e+03],
       [-1.22523898e+04],
       [-1.73772080e+04],
       [ 2.89364032e+02],
       [-1.9

In [37]:
pd.DataFrame((lr_model.coef_).T,index=X_train.columns,\
             columns=['Co-efficients']).sort_values('Co-efficients',ascending=False)

Co-efficients
OverallQual            12490.489322
Fireplaces              9054.341814
OverallCond             7598.427245
Foundation_PConc        5351.383073
BsmtFinType1_GLQ        4885.418935
FullBath                4075.078654
HalfBath                3879.743657
GarageCars              3160.972778
SaleCondition_Normal    2969.797325
BsmtFullBath            2259.531237
TotRmsAbvGrd            1501.552039
YearBuilt                319.863702
ExterCond_TA             289.364032
YearRemodAdd             119.499894
LotFrontage               89.060478
GrLivArea                 28.870016
TotalBsmtSF               26.367312
ScreenPorch               25.370792
MasVnrArea                24.262561
2ndFlrSF                  23.176282
WoodDeckSF                19.709458
GarageArea                17.585789
BsmtFinSF1                15.906086
1stFlrSF                  10.102074
BsmtFinSF2                 9.052728
EnclosedPorch              2.761705
BsmtUnfSF                  1.408498
LotArea                    0.457465
OpenPorchSF               -7.368602
GarageYrBlt              -16.373774
MSSubClass              -153.706581
MoSold                  -436.077770
GarageType_Detchd       -588.016451
YrSold                  -773.112479
Exterior2nd_MetalSd     -863.839218
Exterior2nd_Wd Sdng    -1310.870386
HeatingQC_TA           -1321.063398
BsmtFinType1_Unf       -1556.446922
GarageType_Attchd      -1917.982184
Foundation_CBlock      -1990.704624
GarageFinish_Unf       -2265.777821
GarageFinish_RFn       -2572.594029
HeatingQC_Gd           -2739.622508
Exterior2nd_VinylSd    -3262.306078
Exterior2nd_Plywood    -4381.075590
FireplaceQu_Gd         -4434.826749
Exterior2nd_HdBoard    -4613.875370
BedroomAbvGr           -5162.664103
FireplaceQu_TA         -5663.227989
BsmtExposure_Mn        -6371.669727
BsmtExposure_No        -8409.497811
SaleType_WD           -10623.819837
ExterQual_Gd          -12252.389763
ExterQual_TA          -17377.208021
KitchenQual_TA        -17601.987993
KitchenQual_Gd        -17850.975923
BsmtQual_TA           -18041.766281
BsmtQual_Gd           -21773.506245

#### 7.3 Using the Model for Prediction

In [38]:
y_pred_train = lr_model.predict(X_train)  

In [39]:
y_pred_train                                                         # make predictions on the training set

array([[336305.49494713],
       [ 89339.49266879],
       [122794.01925053],
       ...,
       [155841.80189821],
       [193248.13359668],
       [195909.48807826]])

In [40]:
y_pred_test = lr_model.predict(X_test)                                  # make predictions on the testing set

#### 7.4. Model evaluation using metrics

In [41]:
from sklearn import metrics
MAE_train = metrics.mean_absolute_error(y_train, y_pred_train)
MAE_test = metrics.mean_absolute_error(y_test, y_pred_test)

__Mean Absolute Error__ (MAE) is the mean of the absolute value of the errors:
$$\frac 1n\sum_{i=1}^n|y_i-\hat{y}_i|$$
Computing the MAE for our House Price predictions

In [42]:
print('MAE for training set is {}'.format(MAE_train))
print('MAE for test set is {}'.format(MAE_test))

MAE for training set is 17178.529020777634
MAE for test set is 21852.23520424918


__Mean Squared Error__ (MSE) is the mean of the squared errors:
$$\frac 1n\sum_{i=1}^n(y_i-\hat{y}_i)^2$$

Computing the MSE for our House Price predictions

In [44]:
MSE_train = metrics.mean_squared_error(y_train, y_pred_train)
MSE_test = metrics.mean_squared_error(y_test, y_pred_test)

In [45]:
print('MSE for training set is {}'.format(MSE_train))
print('MSE for test set is {}'.format(MSE_test))

MSE for training set is 761973549.5913721
MSE for test set is 2332597565.765126


__Root Mean Squared Error__ (RMSE) is the square root of the mean of the squared errors:

$$\sqrt{\frac 1n\sum_{i=1}^n(y_i-\hat{y}_i)^2}$$

Computing the RMSE for our House Price predictions

In [47]:
RMSE_train = np.sqrt( metrics.mean_squared_error(y_train, y_pred_train))
RMSE_test = np.sqrt(metrics.mean_squared_error(y_test, y_pred_test))

In [48]:
print('RMSE for training set is {}'.format(RMSE_train))
print('RMSE for test set is {}'.format(RMSE_test))

RMSE for training set is 27603.868380923934
RMSE for test set is 48296.97263561274


#### 7.5 Model Evaluation using Rsquared value.

In [50]:
# r2_score is a standard metric in the sklearn library
r2_train = metrics.r2_score(y_train, y_pred_train)
r2_test = metrics.r2_score(y_test, y_pred_test)
print('R-Squared of train data [Linear Regression]:',r2_train)
print('R-Squared of test data [Linear Regression]:',r2_test)

R-Squared of train data [Linear Regression]: 0.8762419235527997
R-Squared of test data [Linear Regression]: 0.6622287512871425


## 8 Logistic Regression using selected features

## 9 Decision Tree using selected features

### 9.1 Decision Tree in scikit-learn

To apply any machine learning algorithm on your dataset, basically there are 4 steps:
1. Load the algorithm
2. Instantiate and Fit the model to the training dataset
3. Prediction on the test set
4. Evaluate - Calculate RMSE and R square

The code block given below shows how these steps are carried out:<br/>

``` from sklearn import tree
    model = tree.DecisionTreeRegressor(random_state = 0)
    model.fit(X, y)
    predicted= model.predict(x_test)
    ```

In [57]:
from sklearn import tree
modelReg = tree.DecisionTreeRegressor(random_state = 0)
modelReg.fit(X_train, y_train)

DecisionTreeRegressor(random_state=0)

#### 9.1.1 Using the Model for Prediction

In [58]:
y_pred_train = modelReg.predict(X_train)  

In [59]:
y_pred_test = modelReg.predict(X_test)                                                           # make predictions on the testing set

#### 9.1.2 Model evaluation using Metrics

In [61]:
from sklearn import metrics
MAE_train = metrics.mean_absolute_error(y_train, y_pred_train)
MAE_test = metrics.mean_absolute_error(y_test, y_pred_test)

__Mean Absolute Error__ (MAE) is the mean of the absolute value of the errors:
$$\frac 1n\sum_{i=1}^n|y_i-\hat{y}_i|$$
Computing the MAE for our House Price predictions

In [62]:
print('MAE for training set is {}'.format(MAE_train))
print('MAE for test set is {}'.format(MAE_test))

MAE for training set is 7.1061643835616435
MAE for test set is 27522.722602739726


__Mean Squared Error__ (MSE) is the mean of the squared errors:
$$\frac 1n\sum_{i=1}^n(y_i-\hat{y}_i)^2$$

Computing the MSE for our House Price predictions

In [63]:
MSE_train = metrics.mean_squared_error(y_train, y_pred_train)
MSE_test = metrics.mean_squared_error(y_test, y_pred_test)

__Root Mean Squared Error__ (RMSE) is the square root of the mean of the squared errors:

$$\sqrt{\frac 1n\sum_{i=1}^n(y_i-\hat{y}_i)^2}$$

Computing the RMSE for our House Price predictions

In [67]:
RMSE_train = np.sqrt( metrics.mean_squared_error(y_train, y_pred_train))
RMSE_test = np.sqrt(metrics.mean_squared_error(y_test, y_pred_test))

In [68]:
print('RMSE for training set is {}'.format(RMSE_train))
print('RMSE for test set is {}'.format(RMSE_test))

RMSE for training set is 171.72822188499134
RMSE for test set is 52626.16042038851


#### 9.1.3 __R Square__

In [69]:
# r2_score is a standard metric in the sklearn library
r2_train = metrics.r2_score(y_train, y_pred_train)
r2_test = metrics.r2_score(y_test, y_pred_test)
print('R-Squared of train data [Decision Tree]:',r2_train)
print('R-Squared of test data [Decision Tree]:',r2_test)

R-Squared of train data [Decision Tree]: 0.9999952102041766
R-Squared of test data [Decision Tree]: 0.5989613479834592


### 9.2 Decision Tree with Gridsearch

Applying __GridsearchCV__ method for __exhaustive search over specified parameter values__ of estimator.
To know more about the different parameters in decision tree classifier, refer the [documentation](http://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html). <br/>
Below we will apply gridsearch over the following parameters:
- criterion 
- max_depth 
- max_features

You can change other parameters also and compare the impact of it via calculating __MEA MSE RMSE and R Square__

In [74]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import GridSearchCV

decision_tree_regressor = DecisionTreeRegressor(random_state = 0)

tree_para = [{'max_depth': range(2,60),
                             'max_features': ['sqrt', 'log2', None] }]                         #     (What to pass, wht is range )

grid_search = GridSearchCV(decision_tree_regressor,tree_para, cv=10, refit='AUC')
grid_search.fit(X_train, y_train)

GridSearchCV(cv=10, estimator=DecisionTreeRegressor(random_state=0),
             param_grid=[{'max_depth': range(2, 60),
                          'max_features': ['sqrt', 'log2', None]}],
             refit='AUC')

In [75]:
y_pred_train1 = grid_search.predict(X_train)
y_pred_test1 = grid_search.predict(X_test)

#### 9.2.1 Model Evaluation using R Square

In [73]:
# r2_score is a standard metric in the sklearn library
r2_train = metrics.r2_score(y_train, y_pred_train1)
r2_test = metrics.r2_score(y_test, y_pred_test1)
print('R-Squared of train data:',r2_train)
print('R-Squared of test data:',r2_test)

R-Squared of train data: 0.9999952102041766
R-Squared of test data: 0.5989613479834592


### 9.3 Decision Tree with Gridsearch [Hyperparameter Optimization] - Changing Hyperparameters Manually

In [77]:
from sklearn import tree
model_dt_new = tree.DecisionTreeRegressor(max_depth=5,
                       max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=8, min_samples_split=7,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=0, splitter='best')
model_dt_new.fit(X_train, y_train)

DecisionTreeRegressor(max_depth=5, min_samples_leaf=8, min_samples_split=7,
                      presort=False, random_state=0)

In [78]:
y_pred_train_manu = model_dt_new.predict(X_train) 
y_pred_test_manu = model_dt_new.predict(X_test)

#### 9.3.1 Model Evaluation using R Square

In [80]:
# r2_score is a standard metric in the sklearn library
r2_train = metrics.r2_score(y_train, y_pred_train_manu)
r2_test = metrics.r2_score(y_test, y_pred_test_manu)
print('R-Squared of train data [Decision Tree with Gridsearch Changing Hyperparameters Manually]:',r2_train)
print('R-Squared of test data [Decision Tree with Gridsearch Changing Hyperparameters Manually]:',r2_test)

R-Squared of train data [Decision Tree with Gridsearch Changing Hyperparameters Manually]: 0.8376260285446049
R-Squared of test data [Decision Tree with Gridsearch Changing Hyperparameters Manually]: 0.7337676000951491


### 9.4 Hyperparameter Optimization with Gridsearch [Auto Paramater Optimization]

<br/>
Applying __GridsearchCV__ method for __exhaustive search over specified parameter values__ of estimator.
To know more about the different parameters in decision tree classifier, refer the [documentation](http://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html). <br/>
Below we will apply gridsearch over the following parameters:
- criterion 
- max_depth 
- max_features

You can change other parameters also and compare the impact of it via calculating __MAE MSE RMSE R-Square__

In [82]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import GridSearchCV

decision_tree_regressor = DecisionTreeRegressor(random_state = 0)

tree_para = [{'max_depth': range(2,60),
                             'max_features': ['sqrt', 'log2', None] }]
                            
model_dt_tuned = GridSearchCV(decision_tree_regressor,tree_para)

In [83]:
model_dt_tuned.fit(X_train, y_train)

GridSearchCV(estimator=DecisionTreeRegressor(random_state=0),
             param_grid=[{'max_depth': range(2, 60),
                          'max_features': ['sqrt', 'log2', None]}])

In [84]:
model_dt_tuned.best_params_

{'max_depth': 6, 'max_features': None}

In [85]:
model_dt_tuned.scorer_

<function sklearn.metrics._scorer._passthrough_scorer(estimator, *args, **kwargs)>

In [86]:
y_pred_train_tuned1 = model_dt_tuned.predict(X_train)
y_pred_test_tuned1 = model_dt_tuned.predict(X_test)

#### 9.4.1 Model Evaluation using R Square

In [87]:
# r2_score is a standard metric in the sklearn library
r2_train = metrics.r2_score(y_train, y_pred_train_tuned1)
r2_test = metrics.r2_score(y_test, y_pred_test_tuned1)
print('R-Squared of train data:',r2_train)
print('R-Squared of test data:',r2_test)

R-Squared of train data: 0.9094299102723593
R-Squared of test data: 0.6864710690848008


In [88]:
param_grid = [{"criterion": ["mse", "mae"],
              "min_samples_split": [10, 20, 40],
              "max_depth": [2, 6, 8],
              "min_samples_leaf": [20, 40, 100],
              "max_leaf_nodes": [5, 20, 100],
              "max_features": ['sqrt', 'log2', None]
              }]

In [89]:
model_dt_tuned2 = GridSearchCV(decision_tree_regressor,param_grid,cv=10)

In [90]:
model_dt_tuned2.fit(X_train, y_train)

GridSearchCV(cv=10, estimator=DecisionTreeRegressor(random_state=0),
             param_grid=[{'criterion': ['mse', 'mae'], 'max_depth': [2, 6, 8],
                          'max_features': ['sqrt', 'log2', None],
                          'max_leaf_nodes': [5, 20, 100],
                          'min_samples_leaf': [20, 40, 100],
                          'min_samples_split': [10, 20, 40]}])

In [91]:
model_dt_tuned2.best_params_

{'criterion': 'mae',
 'max_depth': 8,
 'max_features': None,
 'max_leaf_nodes': 100,
 'min_samples_leaf': 20,
 'min_samples_split': 10}

In [92]:
model_dt_tuned2.scorer_

<function sklearn.metrics._scorer._passthrough_scorer(estimator, *args, **kwargs)>

In [93]:
y_pred_train_tuned2 = model_dt_tuned2.predict(X_train)
y_pred_test_tuned2 = model_dt_tuned2.predict(X_test)

In [97]:
# r2_score is a standard metric in the sklearn library
r2_train = metrics.r2_score(y_train, y_pred_train_tuned2)
r2_test = metrics.r2_score(y_test, y_pred_test_tuned2)
print('R-Squared of train data [Decision Tree - Auto Paramater Optimization]:',r2_train)
print('R-Squared of test data [Decision Tree - Auto Paramater Optimization]:',r2_test)

R-Squared of train data [Decision Tree - Auto Paramater Optimization]: 0.8340658494396264
R-Squared of test data [Decision Tree - Auto Paramater Optimization]: 0.7555613562970959


## 10 Random Forest using selected features

## 10.1 Random Forest without parameter in scikit-learn

To apply any machine learning algorithm on your dataset, basically there are 4 steps:
1. Load the algorithm
2. Instantiate and Fit the model to the training dataset
3. Prediction on the test set
4. Calculating the accuracy of the model

The code block given below shows how these steps are carried out:<br/>

``` from sklearn.ensemble import RandomForestRegressor
    model = RandomForestRegressor()
    model.fit(X, y)
    predicted= model.predict(x_test)
    ```

#### 10.1.1 __Model without parameter specification__

In [100]:
from sklearn.ensemble import RandomForestRegressor

model_rf = RandomForestRegressor(random_state = 0)

model_rf.fit(X_train, y_train)

RandomForestRegressor(random_state=0)

#### 10.1.2 Using the Model for Prediction

In [102]:
y_pred_train_rf = model_rf.predict(X_train) 
y_pred_test_rf = model_rf.predict(X_test) 

#### 10.1.3 Model evaluation for Random Forest

In [105]:
# r2_score is a standard metric in the sklearn library
r2_train = metrics.r2_score(y_train, y_pred_train_rf)
r2_test = metrics.r2_score(y_test, y_pred_test_rf)
print('R-Squared of train data [Random Forest]:',r2_train)
print('R-Squared of test data [Random Forest]:',r2_test)


R-Squared of train data [Random Forest]: 0.9795352102052173
R-Squared of test data [Random Forest]: 0.8254164949787026


## 10.2 Random Forest - Hyper parameter for RF tuning using RANDOM SEARCH

In [107]:
from sklearn.model_selection import RandomizedSearchCV
# parameters for RandomSearchCV
param_dist = {"max_depth": range(2,6),
              "min_samples_split": range(6, 11),
              "min_samples_leaf": range(1, 5),
              "bootstrap": [True,False],
              "n_estimators": [100, 400],
              'max_features': ['sqrt', 'log2',5]
             }
# run randomized search
n_iter_search = 100
model_rf_tuned = RandomizedSearchCV(model_rf, param_distributions = param_dist,
                                   n_iter = n_iter_search,
                                   n_jobs = -1)

In [108]:
model_rf_tuned.fit(X_train, y_train)

RandomizedSearchCV(estimator=RandomForestRegressor(random_state=0), n_iter=100,
                   n_jobs=-1,
                   param_distributions={'bootstrap': [True, False],
                                        'max_depth': range(2, 6),
                                        'max_features': ['sqrt', 'log2', 5],
                                        'min_samples_leaf': range(1, 5),
                                        'min_samples_split': range(6, 11),
                                        'n_estimators': [100, 400]})

In [111]:
model_rf_tuned.best_params_

{'n_estimators': 400,
 'min_samples_split': 9,
 'min_samples_leaf': 3,
 'max_features': 'sqrt',
 'max_depth': 5,
 'bootstrap': True}

- Using the model for __prediction__

In [112]:
y_pred_train_rf_tuned = model_rf_tuned.predict(X_train)
y_pred_test_rf_tuned = model_rf_tuned.predict(X_test)

#### 10.2.1 - __Model Evaluation__ using __r2_score__

In [114]:
# r2_score is a standard metric in the sklearn library
r2_train = metrics.r2_score(y_train, y_pred_train_rf_tuned)
r2_test = metrics.r2_score(y_test, y_pred_test_rf_tuned)
print('R-Squared of train data [Hyper parameter for Random Forest tuning using RANDOM SEARCH ]:',r2_train)
print('R-Squared of test data [Hyper parameter for Random Forest tuning using RANDOM SEARCH ]:',r2_test)

R-Squared of train data [Hyper parameter for Random Forest tuning using RANDOM SEARCH ]: 0.8726930439127516
R-Squared of test data [Hyper parameter for Random Forest tuning using RANDOM SEARCH ]: 0.7919625630531667


## 11 Comparision of Models

### 11.1 Model - Linear Regression

R-Squared of train data [Linear Regression]: 0.8762419235527997 <BR>
R-Squared of test data [Linear Regression]: 0.6622287512871425

### 11.2 Model - Logistic Regression

### 11.3 Model - Decision Tree

R-Squared of train data [Decision Tree - Auto Paramater Optimization]: 0.8340658494396264 <br>
R-Squared of test data [Decision Tree - Auto Paramater Optimization]: 0.7555613562970959

### 12.4 Model - Random Forest

R-Squared of train data [Hyper parameter for Random Forest tuning using RANDOM SEARCH ]: 0.8726930439127516 <BR>
R-Squared of test data [Hyper parameter for Random Forest tuning using RANDOM SEARCH ]: 0.7919625630531667